composectrl@kenit.cc/files/composectrl@kenit.cc/applet.js

```diff
@@ -0,0 +1,190 @@
+            });
+            return new Promise((resolve, reject) => {
+                let out_cancel = new Gio.Cancellable();
+                let err_cancel = new Gio.Cancellable();
+                let cb = (reader, res)=>{
+                    let [out, length] = reader.read_upto_finish(res);
+                    let result = out.toString().replace(/\W/g, '');
+                    if(result == 'ok'){                        
+                        resolve(result);
+                        err_cancel.cancel();
+                    }else{
+                        reject(result);
+                        out_reader.cancel();
+                    }
+                };
+                out_reader.read_upto_async("", 0, 0, out_cancel, cb, "");
+                err_reader.read_upto_async("", 0, 0, err_cancel, cb, "");
+            });            
+        })).then(() => {
+            this._set_applet_status(true);
+        }).catch(reason => {
+            this._set_applet_status(false);
+        });
+    },
+    _set_applet_status: function(status){
+        let icon_name = (status ? 'Running' : 'Stopped');
+        this.set_applet_icon_path(this._path + '/icons/128/' + icon_name + '.png');
+        this.set_applet_tooltip(_(status ? "All services are running" : "Partial or all services are down"));
+    },
+    _setMenu: function(){
+        let upItem = new PopupMenu.PopupMenuItem(_("Up"));
+        upItem.connect('activate', () => this._up());
+        this.menu.addMenuItem(upItem);
+        
+        let downItem = new PopupMenu.PopupMenuItem(_("Down"));
+        downItem.connect('activate', () => this._down());
+        this.menu.addMenuItem(downItem);
+
+        let restartItem = new PopupMenu.PopupMenuItem(_("Restart"));
+        restartItem.connect('activate', () => this._restart());
+        this.menu.addMenuItem(restartItem);
+
+        let refreshItem = new PopupMenu.PopupMenuItem(_("Refresh"));
+        refreshItem.connect('activate', () => this._check());
+        this.menu.addMenuItem(refreshItem);
+    },
+    _down: function(){
+        Util.spawn_async([this.composeCmd, '-p', this.composeProjectName, '-f', this.composeFilePath, 'down'], () => this._get_service_status());
+    },
+    _up: function(){
+        Util.spawn_async([this.composeCmd, '-p', this.composeProjectName, '-f', this.composeFilePath, 'up', '-d'], () => this._get_service_status());
+    },
+    _restart: function(){
+        Util.spawn_async([this.composeCmd, '-p', this.composeProjectName, '-f', this.composeFilePath, 'restart'], () => this._get_service_status());
+    },
+    _check: function(){
+        this._get_service_status();
+    },
+    _setServiceList: function(){
+        let cmd = [this.composeCmd, '-p', this.composeProjectName, '-f', this.composeFilePath, "config", "--service"];
+
+        let [res, pid, in_fd, out_fd, err_fd]  = GLib.spawn_async_with_pipes(null, cmd, null, GLib.SpawnFlags.SEARCH_PATH, null);
```

Comment:
I suggest adding a try-catch here and notifying the user if they need docker-compose installed, the applet just crashes.

```
Cjs-Message: JS LOG: Failed to execute child process "docker-compose" (No such file or directory)
Cjs-Message: JS LOG: 
<----------------
anonymous/MyApplet.prototype._setServiceList@/home/jason/.local/share/cinnamon/applets/composectrl@kenit.cc/applet.js:156:50
anonymous/MyApplet.prototype._init@/home/jason/.local/share/cinnamon/applets/composectrl@kenit.cc/applet.js:27:9
MyApplet@/home/jason/.local/share/cinnamon/applets/composectrl@kenit.cc/applet.js:9:5
main@/home/jason/.local/share/cinnamon/applets/composectrl@kenit.cc/applet.js:189:12
createApplet@/usr/share/cinnamon/js/ui/appletManager.js:563:18
addAppletToPanels@/usr/share/cinnamon/js/ui/appletManager.js:341:22
finishExtensionLoad@/usr/share/cinnamon/js/ui/appletManager.js:97:14
Extension.prototype._init/<@/usr/share/cinnamon/js/ui/extension.js:250:17
---------------->
Cjs-Message: JS LOG: 
[composectrl@kenit.cc]: Applet composectrl@kenit.cc: Could not create applet object.
[composectrl@kenit.cc]: Error importing applet.js from composectrl@kenit.cc
Cjs-Message: JS LOG: 
<----------------
Extension.prototype._init/<@/usr/share/cinnamon/js/ui/extension.js:251:1
---------------->
```